fundamental-concepts/contract-keys/4_using_contract_keys.md

```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
 
 <pre class="file" data-filename="daml/User.daml" data-target="append">
-  (aliceCid1, aliceUser) <- submit alice $ fetchByKey @User alice
-  assert $ aliceUser == User with username=alice, following=[bob]
-  return ()
-</pre>
 
-- `fetchByKey` takes as first argument the template name, here `@User`, and as a second argument
-  the contract key
-- it returns a tuple, where the first factor is the actual contract ID and the second factor is the
-  contract data
+  userContractIdAndContract <- queryContractKey @User alice (alice)
+  -- We get back an Optional Tuple, like in Haskell we can
+  -- first use fromSome to unpack the Optional Tuple when it is Some Tuple
+  -- then we access the first and second elements with helper functions (fst, snd),
+  -- or alternatively access them positionally with t._1, t._2 and so on
 
-`lookupByKey` is similar to `fetchByKey`, but instead of failing the transaction, it will returns a
-`None` if the given key does not exist.
+  let contractId = fst $ fromSome userContractIdAndContract
+  let contract = snd $ fromSome userContractIdAndContract
+  assert $ contract == User with username=alice, following=[bob]
+  assert $ aliceContractId == contractId
+  return()
+</pre>
 
-Remove the last line and append
+- `queryContractKey` takes as first argument the `template` name (`@User`), the `Party` that will make the query as the second argument (`alice`), and the contract key as the third argument (`(alice)`)
+- `fetchByKey` takes only two argument, i.e., the template name and the contract key
+- Both actions return a tuple, where the first factor is the actual contract ID and the second factor is the
+  contract data
 
-<pre class="file" data-filename="daml/User.daml" data-target="append">
-  Some aliceCid2 <- submit alice $ lookupByKey @User alice
-  assert $ aliceCid1 == aliceCid2
-  return ()
-</pre>
+Since we didn't exercise a choice on `Alice` `User` contract, we expect the returned contract ID
+to be unchanged. We check if this is true with `assert $ aliceContractId == contractId`.
 
-- The first argument to `lookupByKey` is the template name and the second the contract key
-- Since we didn't exercise an choice on `Alice` `User` contract, we expect the returned contract ID
-to be unchanged
+As Daml Script offers only `queryContractKey` for both `fetchByKey` and `lookupByKey` we will explain the similarities and differences between the two:
 
-`lookupByKey` does not give any information of whether a contract with a given key exists or not.
+- `lookupByKey` takes the same two arguments as `fetchByKey`, i.e., the template name and the contract key
+- `lookupByKey` is similar to `fetchByKey`, but instead of failing the transaction, it will return a
+`None` if the given key does not exist.
+- `lookupByKey` does not give any information of whether a contract with a given key exists or not.
```

Comment:
I think this line is wrong. It's `fetchByKey` that does not give any information about existence of a contract.

Reply:
I took this from the original Katacoda scenario, you can find it [here on step 4](https://beta.katacoda.com/daml/courses/fundamental-concepts/contract-keys).

> lookupByKey does not give any information of whether a contract with a given key exists or not.

Reply:
Just checked the docs, you're right. `fetchByKey` doesn't disclose any info on contract's existence.